File: typescript/processor/src/output/serializer.ts
// @docspec:module {
//   id: "docspec-ts-serializer",
//   name: "DocSpec Output Serializer",
//   description: "Serializes the final DocSpec output to docspec.json and optionally intent-graph.json. Creates the output directory if it does not exist and writes formatted JSON files.",
//   since: "3.0.0"
// }

import * as fs from "node:fs";
import * as path from "node:path";

/**
 * @docspec:boundary "File system output for docspec.json and intent-graph.json"
 */
export class Serializer {
  /** @docspec:intentional "Writes docspec.json and optionally intent-graph.json to the output directory" */
  write(data: any, outputDir: string): void {
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    // Write docspec.json
    const specPath = path.join(outputDir, "docspec.json");
    fs.writeFileSync(specPath, JSON.stringify(data, null, 2), "utf-8");

    // Write intent-graph.json if present
    if (data.intentGraph) {
      const intentPath = path.join(outputDir, "intent-graph.json");
      const intentData = {
        docspec: data.docspec,
        artifact: { groupId: data.artifact.groupId, artifactId: data.artifact.artifactId, version: data.artifact.version },
        methods: data.intentGraph.methods,
      };
      fs.writeFileSync(intentPath, JSON.stringify(intentData, null, 2), "utf-8");
    }
  }
}

File: packages/processor/python/docspec_processor/output/serializer.py
"""Write docspec.json output."""
# @docspec:module {
#   id: "docspec-py-serializer",
#   name: "Spec Serializer",
#   description: "Serializes the DocSpec output model to docspec.json and intent-graph.json files, stripping None values during serialization.",
#   since: "3.0.0"
# }
import json
import os
from pathlib import Path


# @docspec:boundary "JSON output serialization for docspec.json and intent-graph.json"
class Serializer:
    # @docspec:method { since: "3.0.0" }
    # @docspec:intentional "Serialize the output model to docspec.json and optionally intent-graph.json"
    def write(self, data: dict, output_dir: str) -> None:
        os.makedirs(output_dir, exist_ok=True)
        spec_path = Path(output_dir) / "docspec.json"
        # Remove None values
        cleaned = self._clean(data)
        with open(spec_path, "w", encoding="utf-8") as f:
            json.dump(cleaned, f, indent=2, ensure_ascii=False)

        # Write intent-graph.json if present
        if data.get("intentGraph"):
            intent_path = Path(output_dir) / "intent-graph.json"
            intent_data = {
                "docspec": data["docspec"],
                "artifact": {
                    "groupId": data["artifact"]["groupId"],
                    "artifactId": data["artifact"]["artifactId"],
                    "version": data["artifact"]["version"],
                },
                "methods": data["intentGraph"]["methods"],
            }
            with open(intent_path, "w", encoding="utf-8") as f:
                json.dump(self._clean(intent_data), f, indent=2, ensure_ascii=False)

    # @docspec:deterministic
    # @docspec:intentional "Recursively strip None values from nested dicts and lists"
    def _clean(self, obj):
        if isinstance(obj, dict):
            return {k: self._clean(v) for k, v in obj.items() if v is not None}
        elif isinstance(obj, list):
            return [self._clean(item) for item in obj]
        return obj
